src/components/swap/ConfirmSwapModal.tsx

```diff
@@ -1,39 +1,172 @@
+import { ConfirmationModalContent } from '../TransactionConfirmationModal'
+import {
+  ErrorModalContent,
+  PendingConfirmModalState,
+  PendingModalContent,
+  PendingModalError,
+} from './PendingModalContent'
 import SwapModalFooter from './SwapModalFooter'
 import SwapModalHeader from './SwapModalHeader'
 
+export enum ConfirmModalState {
+  REVIEWING,
+  APPROVING_TOKEN,
+  PERMITTING,
+  PENDING_CONFIRMATION,
+}
+
+function isAllowing(confirmModalState: ConfirmModalState) {
+  return confirmModalState === ConfirmModalState.APPROVING_TOKEN || confirmModalState === ConfirmModalState.PERMITTING
+}
+
+function useConfirmModalState({
+  trade,
+  allowedSlippage,
+  onSwap,
+  allowance,
+  doesTradeDiffer,
+}: {
+  trade: InterfaceTrade
+  allowedSlippage: Percent
+  onSwap: () => void
+  allowance: Allowance
+  doesTradeDiffer: boolean
+}) {
+  const [confirmModalState, setConfirmModalState] = useState<ConfirmModalState>(ConfirmModalState.REVIEWING)
+  const [approvalError, setApprovalError] = useState<PendingModalError>()
+  const [pendingModalSteps, setPendingModalSteps] = useState<PendingConfirmModalState[]>([])
+
+  const prepareSwapFlow = useCallback(() => {
+    const steps: PendingConfirmModalState[] = []
+    if (allowance.state === AllowanceState.REQUIRED && allowance.needsPermit2Approval) {
+      steps.push(ConfirmModalState.APPROVING_TOKEN)
+    }
+    if (allowance.state === AllowanceState.REQUIRED && allowance.needsSignature) {
+      steps.push(ConfirmModalState.PERMITTING)
+    }
+    steps.push(ConfirmModalState.PENDING_CONFIRMATION)
+    setPendingModalSteps(steps)
+  }, [allowance])
+
+  const { chainId } = useWeb3React()
+  const trace = useTrace()
+  const maximumAmountIn = useMaxAmountIn(trade, allowedSlippage)
+  // Starts the approval process, by triggering either the Token Approval or the Permit signature.
+  const updateAllowance = useCallback(async () => {
+    invariant(allowance.state === AllowanceState.REQUIRED)
+    try {
+      if (allowance.needsPermit2Approval) {
+        setConfirmModalState(ConfirmModalState.APPROVING_TOKEN)
+        await allowance.approve()
+        sendAnalyticsEvent(InterfaceEventName.APPROVE_TOKEN_TXN_SUBMITTED, {
+          chain_id: chainId,
+          token_symbol: maximumAmountIn?.currency.symbol,
+          token_address: maximumAmountIn?.currency.address,
+          ...trace,
+        })
+      } else {
+        setConfirmModalState(ConfirmModalState.PERMITTING)
+        await allowance.permit()
+      }
+    } catch (e) {
+      setConfirmModalState(ConfirmModalState.REVIEWING)
+      if (didUserReject(e)) {
+        return
+      }
+      console.error(e)
+      setApprovalError(
+        allowance.needsPermit2Approval ? PendingModalError.TOKEN_APPROVAL_ERROR : PendingModalError.PERMIT_ERROR
+      )
+    }
+  }, [allowance, chainId, maximumAmountIn?.currency.address, maximumAmountIn?.currency.symbol, trace])
+
+  const onStartSwapFlow = useCallback(async () => {
```

Comment:
i feel like one big change i'd prefer is, instead of having the main modal control updating the state for every screen based on the combination of these callbacks and useEffects, it would be better for each child screen to control its own `confirm` method and give it a `nextStep()` callback they can call when it's done. lmk if this makes sense, i can provide a code snippet if not

Reply:
Do we have separate child screens tho? The problem, which I outlined in the other PR is that we have a dynamic Modal component and we just replace the bottomContent/content callbacks.

I think this is slightly cleaner, at least we have unified control flow. I fear that if we spread it across multiple components, we may loose a sight of what's going on in an already quite complex file.